src/RwaUrn.sol

```diff
@@ -0,0 +1,125 @@
+    modifier auth {
+        require(wards[msg.sender] == 1, "RwaUrn/not-authorized");
+        _;
+    }
+    function hope(address usr) external auth {
+        can[usr] = 1;
+        emit Hope(usr);
+    }
+    function nope(address usr) external auth {
+        can[usr] = 0;
+        emit Nope(usr);
+    }
+    modifier operator {
+        require(can[msg.sender] == 1, "RwaUrn/not-operator");
+        _;
+    }
+
+    VatAbstract  public vat;
+    GemJoinAbstract public gemJoin;
+    DaiJoinAbstract public daiJoin;
+    address public outputConduit;
+
+    // Events
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Hope(address indexed usr);
+    event Nope(address indexed usr);
+    event File(bytes32 indexed what, address data);
+    event Lock(uint256 wad);
+    event Free(uint256 wad);
+    event Draw(uint256 wad);
+    event Wipe(uint256 wad);
+
+    // --- init ---
+    constructor(
+        address vat_, address gemJoin_, address daiJoin_, address outputConduit_
+    ) public {
+        // requires in urn that outputConduit isn't address(0)
+        vat = VatAbstract(vat_);
+        gemJoin = GemJoinAbstract(gemJoin_);
+        daiJoin = DaiJoinAbstract(daiJoin_);
+        outputConduit = outputConduit_;
+        wards[msg.sender] = 1;
+        DSTokenAbstract(gemJoin.gem()).approve(address(gemJoin), uint256(-1));
+        DaiAbstract(daiJoin.dai()).approve(address(daiJoin), uint256(-1));
+        VatAbstract(vat_).hope(address(daiJoin));
+        emit Rely(msg.sender);
+    }
+
+    // --- administration ---
+    function file(bytes32 what, address data) external auth {
+        if (what == "outputConduit") { outputConduit = data; }
+        else revert("RwaUrn/unrecognised-param");
+        emit File(what, data);
+    }
+
+    // --- cdp operation ---
+    // n.b. that the operator must bring the gem
+    function lock(uint256 wad) external operator {
+        require(wad <= 2**255 - 1, "RwaUrn/overflow");
+        DSTokenAbstract(gemJoin.gem()).transferFrom(msg.sender, address(this), wad);
+        // join with address this
+        gemJoin.join(address(this), wad);
+        vat.frob(gemJoin.ilk(), address(this), address(this), address(this), int(wad), 0);
+        emit Lock(wad);
+    }
+    // n.b. that the operator takes the gem
+    // and might not be the same operator who brought the gem
+    function free(uint256 wad) external operator {
+        require(wad <= 2**255, "RwaUrn/overflow");
+        vat.frob(gemJoin.ilk(), address(this), address(this), address(this), -int(wad), 0);
+        gemJoin.exit(msg.sender, wad);
+        emit Free(wad);
+    }
+    // n.b. DAI can only go to the output conduit
+    function draw(uint256 wad) external operator {
+        require(outputConduit != address(0));
+        require(wad <= 2**255 - 1, "RwaUrn/overflow");
+        vat.frob(gemJoin.ilk(), address(this), address(this), address(this), 0, int(wad));
+        daiJoin.exit(outputConduit, wad);
```

Comment:
`wad` is not the same for these two lines. `dart` in frob is a normalized `art` whereas `daiJoin.exit()` expects the actual dai amount. This will not fail until SFs are accumulated, so a test should be added for this case.

Reply:
Great spot 👍

Reply:
Yeah, I'm thinking we should always test frobs with a time delay and drip() before a draw.

Reply:
I've pushed a commit fixing this and adding a basic fee accumulation to the tests. Fee accumulation should probably be tested more, including:
- time delay and `drip()` before first draw
- full repayment after interest accumulation